udata_recommendations/views.py

```diff
@@ -8,14 +8,33 @@
 blueprint = Blueprint('recommendations', __name__, template_folder='templates')
 
 
+NB_RECOMMENDATIONS = 2
+
+
 def has_recommendations(ctx):
     dataset = ctx['dataset']
-    return 'recommendations' in dataset.extras and len(dataset.extras['recommendations'])
+    if not dataset:
+        return False
+    recommendations = dataset.extras.get('recommendations', [])
+    return recommendations and len(recommendations) > 0
 
 
 @template_hook('dataset.display.after-description', when=has_recommendations)
 def dataset_recommendations(ctx):
-    reco_ids = ctx['dataset'].extras['recommendations']
-    recommendations = Dataset.objects.filter(id__in=reco_ids)
-    return theme.render('dataset-recommendations.html',
-                        recommendations=recommendations)
+    reco_ids = []
+    recommendations = ctx['dataset'].extras['recommendations']
+
+    # Get at most n unique recommendations
+    # Recommendations are already sorted by score in desc order
+    while len(reco_ids) < current_app.config.get('RECOMMENDATIONS_NB_RECOMMENDATIONS'):
```

Comment:
```suggestion
    reco_ids = list({r['id'] for r in recommendations})[:RECOMMENDATIONS_NB_RECOMMENDATIONS]
```

This might be more Perl than Python 🙈 

`{}` is a set comprehension, which should handle deduplication (and keep the best score only for a given id). We need to cast to list to slice the set (and slicing won't fail if they're not enough elements).

Reply:
Almost there! sets aren't ordered so this code doesn't pass the tests. You need to use a dict instead of a set.

https://stackoverflow.com/questions/51145707/using-ordered-dictionary-as-ordered-set